src/rp2_common/hardware_i2c/i2c.c

```diff
@@ -75,11 +75,22 @@ uint i2c_set_baudrate(i2c_inst_t *i2c, uint baudrate) {
     invalid_params_if(I2C, hcnt < 8);
     invalid_params_if(I2C, lcnt < 8);
 
-    // The SDA hold time should be at least 300 ns, per the I2C standard.
-    // sda_tx_hold_count [cycles] = freq_in [cycles/s] * 300ns * (1s / 1e9ns)
-    // Reduce 300/1e9 to 3/1e7 to avoid numbers that don't fit in uint.
-    // Add 1 to avoid division truncation.
-    uint sda_tx_hold_count = ((freq_in * 3) / 10000000) + 1;
+    // Per I2C-bus specification a device in standard or fast mode must
+    // internally provide a hold time of at least 300ns for the SDA signal to
+    // bridge the undefined region of the falling edge of SCL. A smaller hold
+    // time of 120ns is used for fast mode plus.
+    uint sda_tx_hold_count;
+    if (baudrate < 1000000) {
+        // sda_tx_hold_count = freq_in [cycles/s] * 300ns * (1s / 1e9ns)
+        // Reduce 300/1e9 to 3/1e7 to avoid numbers that don't fit in uint.
+        // Add 1 to avoid division truncation.
+        sda_tx_hold_count = ((freq_in * 3) / 10000000) + 1;
+    } else {
+        // sda_tx_hold_count = freq_in [cycles/s] * 120ns * (1s / 1e9ns)
+        // Reduce 120/1e9 to 12/1e8 to avoid numbers that don't fit in uint.
+        // Add 1 to avoid division truncation.
+        sda_tx_hold_count = ((freq_in * 12) / 100000000) + 1;
```

Comment:
You reduce this to the smallest possible integer numerator and denominator here.  `3` and `10000000` for the non-1MHz case, and `3` and `25000000` for 1MHz.

Reply:
```c
uint sda_tx_hold_count = ((freq_in * 3) / (baudrate < 1000000 ? 10000000 : 25000000)) + 1;
```

Reply:
Good idea. 120/1e9 reduced to 3/25e6. I choose readability over compactness here.